crates/heaptrack/src/bpf.rs

```diff
@@ -0,0 +1,274 @@
+
+    pub fn attach_free(&mut self, libc_path: &Path) -> Result<()> {
+        let free_opts = UprobeOpts {
+            func_name: Some("free".to_string()),
+            retprobe: false,
+            ..Default::default()
+        };
+
+        let link = self
+            .skel
+            .progs
+            .uprobe_free
+            .attach_uprobe_with_opts(-1, libc_path, 0, free_opts)
+            .context(format!(
+                "Failed to attach free uprobe in {}",
+                libc_path.display()
+            ))?;
+        self.probes.push(link);
+
+        Ok(())
+    }
+
+    pub fn attach_calloc(&mut self, libc_path: &Path) -> Result<()> {
+        let calloc_opts = UprobeOpts {
+            func_name: Some("calloc".to_string()),
+            retprobe: false,
+            ..Default::default()
+        };
+
+        let link = self
+            .skel
+            .progs
+            .uprobe_calloc
+            .attach_uprobe_with_opts(-1, libc_path, 0, calloc_opts)
+            .context(format!(
+                "Failed to attach calloc uprobe in {}",
+                libc_path.display()
+            ))?;
+        self.probes.push(link);
+
+        // Attach calloc return
+        let calloc_ret_opts = UprobeOpts {
+            func_name: Some("calloc".to_string()),
+            retprobe: true,
+            ..Default::default()
+        };
+        let link = self
+            .skel
+            .progs
+            .uretprobe_calloc
+            .attach_uprobe_with_opts(-1, libc_path, 0, calloc_ret_opts)
+            .context(format!(
+                "Failed to attach calloc uretprobe in {}",
+                libc_path.display()
+            ))?;
+        self.probes.push(link);
+
+        Ok(())
+    }
+
+    pub fn attach_realloc(&mut self, libc_path: &Path) -> Result<()> {
+        let realloc_opts = UprobeOpts {
+            func_name: Some("realloc".to_string()),
+            retprobe: false,
+            ..Default::default()
+        };
+
+        let link = self
+            .skel
+            .progs
+            .uprobe_realloc
+            .attach_uprobe_with_opts(-1, libc_path, 0, realloc_opts)
+            .context(format!(
+                "Failed to attach realloc uprobe in {}",
+                libc_path.display()
+            ))?;
+        self.probes.push(link);
+
+        // Attach realloc return
+        let realloc_ret_opts = UprobeOpts {
+            func_name: Some("realloc".to_string()),
+            retprobe: true,
+            ..Default::default()
+        };
+        let link = self
+            .skel
+            .progs
+            .uretprobe_realloc
+            .attach_uprobe_with_opts(-1, libc_path, 0, realloc_ret_opts)
+            .context(format!(
+                "Failed to attach realloc uretprobe in {}",
+                libc_path.display()
+            ))?;
+        self.probes.push(link);
+
+        Ok(())
+    }
+
+    pub fn attach_aligned_alloc(&mut self, libc_path: &Path) -> Result<()> {
```

Comment:
overall there is a lot of repetition in this file. Can't we extract the key parts for each function we want to attach and factorize the common code ?